arch/arm64/arm64_litecov.cpp

```diff
@@ -0,0 +1,278 @@
+
+  // 3 instr
+  arm64asm->ReadRegStack(module, X1, 8);
+  arm64asm->ReadRegStack(module, X0, 0);
+  arm64asm->OffsetStack(module, 16);
+
+  if (sp_offset) {
+    arm64asm->OffsetStack(module, sp_offset);
+  }
+}
+
+InstructionResult LiteCov::InstrumentInstruction(ModuleInfo *module,
+                                                 Instruction &inst,
+                                                 size_t bb_address,
+                                                 size_t instruction_address) {
+  if (!compare_coverage) {
+    return INST_NOTHANDLED;
+  }
+
+  if (inst.instr.opcode != arm64::Opcode::kSubImmediate &&
+      inst.instr.opcode != arm64::Opcode::kAddImmediate &&
+      inst.instr.opcode != arm64::Opcode::kSubShiftedRegister &&
+      inst.instr.opcode != arm64::Opcode::kAddShiftedRegister &&
+      inst.instr.opcode != arm64::Opcode::kSubExtendedRegister &&
+      inst.instr.opcode != arm64::Opcode::kAddExtendedRegister) {
+    return INST_NOTHANDLED;
+  }
+
+  // cmp instructions use the zero register as destination
+  arm64::Register rd = std::get<arm64::Register>(inst.instr.operands[0]);
+  if (rd.name != arm64::Register::kXzr) {
+    return INST_NOTHANDLED;
+  }
+
+  int operand_width = rd.size;
+
+  // kSubExtendedRegister / kAddExtendedRegister might have sizes smaller than
+  // 32 bit. In case of 8 bit cmp, skip, in case of 16bit cmp add and-instr
+  // after xor.
+  bool is_ext_16_bit = false;
+
+  if (inst.instr.opcode == arm64::Opcode::kAddExtendedRegister ||
+      inst.instr.opcode == arm64::Opcode::kSubExtendedRegister) {
+    arm64::Extend extend = std::get<arm64::Extend>(inst.instr.operands[3]);
+
+    // skip byte cmp
+    if (extend.type == arm64::Extend::Type::kUxtb ||
+        extend.type == arm64::Extend::Type::kSxtb) {
+      return INST_NOTHANDLED;
+    }
+  }
+
+  arm64::Register rn = std::get<arm64::Register>(inst.instr.operands[1]);
+
+  ModuleCovData *data = (ModuleCovData *)module->client_data;
+
+  size_t bb_offset = bb_address - module->min_address;
+  size_t cmp_offset = instruction_address - bb_address;
+  if (cmp_offset >= 0x1000000) {
+    // only allow one cmp instrumentation per bb
+    WARN("Too large basic block for cmp coverage\n");
+    return INST_NOTHANDLED;
+  }
+
+  // check what we matched already
+  int match_width = operand_width - 8;
+  for (; match_width >= 8; match_width -= 8) {
+    uint64_t already_matched_code =
+        GetCmpCode(bb_offset, cmp_offset, match_width);
+    if (data->ignore_coverage.find(already_matched_code) !=
+        data->ignore_coverage.end()) {
+      break;
+    }
+  }
+  match_width += 8;
+  if (match_width == operand_width) {
+    // we already have an (almost) full match
+    return INST_NOTHANDLED;
+  }
+
+  size_t instrumentation_start_offset = module->instrumented_code_allocated;
+
+  Arm64Assembler *arm64asm = static_cast<Arm64Assembler *>(assembler_);
+
+  // start with NOP that's going to be replaced with
```

Comment:
I wonder if it's needed here. On Intel, instructions can be different size, so my concern that writing jump will replace the first but also the part of the second instruction, which might then corrupt that instruction if some thread is right at that point (between the 1st and 2nd instruction). However on ARM all instruction are of the same size which I think eliminates that concern.

Reply:
I included the NOP before for easier debugging, but you are right it is not really needed -> removed it.